subsys/net/lib/sockets/sockets_can.c

```diff
@@ -82,32 +93,76 @@ static void zcan_received_cb(struct net_context *ctx, struct net_pkt *pkt,
 			     union net_proto_header *proto_hdr,
 			     int status, void *user_data)
 {
-	NET_DBG("ctx %p pkt %p st %d ud %p", ctx, pkt, status, user_data);
+	/* The ctx parameter is not really relevant here. It refers to first
+	 * net_context that was used when registering CAN socket.
+	 * In practice there can be multiple sockets that are interested in
+	 * same CAN id packets. That is why we need to implement the dispatcher
+	 * which will give the packet to correct net_context(s).
+	 */
+	struct net_pkt *clone = NULL;
+	int i;
 
-	/* if pkt is NULL, EOF */
-	if (!pkt) {
-		struct net_pkt *last_pkt = k_fifo_peek_tail(&ctx->recv_q);
+	for (i = 0; i < ARRAY_SIZE(receivers); i++) {
+		struct can_frame *frame = (struct can_frame *)
+							pkt->buffer->data;
+		canid_t can_id;
 
-		if (!last_pkt) {
-			/* If there're no packets in the queue, recv() may
-			 * be blocked waiting on it to become non-empty,
-			 * so cancel that wait.
+		if (receivers[i].iface != net_pkt_iface(pkt)) {
+			continue;
+		}
+
+		can_id = UNALIGNED_GET(&frame->can_id);
+
+		if ((can_id & receivers[i].can_mask) !=
+		    (receivers[i].can_id & receivers[i].can_mask)) {
+			continue;
+		}
+
+		if (pkt != NULL && ARRAY_SIZE(receivers) > 1) {
+			/* There are multiple receivers, we need to clone
+			 * the packet.
 			 */
-			sock_set_eof(ctx);
-			k_fifo_cancel_wait(&ctx->recv_q);
-			NET_DBG("Marked socket %p as peer-closed", ctx);
+			clone = net_pkt_clone_new(pkt, MEM_ALLOC_TIMEOUT);
 		} else {
-			net_pkt_set_eof(last_pkt, true);
-			NET_DBG("Set EOF flag on pkt %p", ctx);
+			clone = pkt;
 		}
 
-		return;
-	}
+		ctx = receivers[i].ctx;
+
+		NET_DBG("[%d] ctx %p pkt %p st %d", i, ctx, clone, status);
 
-	/* Normal packet */
-	net_pkt_set_eof(pkt, false);
+		/* if pkt is NULL, EOF */
+		if (!clone) {
```

Comment:
Doesn't exist an error code for out of memory?

Reply:
Yes, that needs to be checked.